python/GafferImageTest/ImageReaderTest.py

```diff
@@ -298,8 +298,10 @@ def testFrameRangeMask( self ) :
 		# so we can tell if BlackOutside is working.
 		blackTile = IECore.FloatVectorData( [ 0 ] * GafferImage.ImagePlug.tileSize() * GafferImage.ImagePlug.tileSize() )
 		with context :
+			self.assertEqual( reader["availableFrames"].getValue(), IECore.IntVectorData( [ 1, 3 , 5, 7 ] ) )
 			for i in range( 1, 11 ) :
 				context.setFrame( i )
+				self.assertEqual( reader["fileValid"].getValue(), i in ( 1, 3, 5, 7 ), "frame {}".format( i ) )
 				self.assertNotEqual( reader["out"].channelData( "R", imath.V2i( 0 ) ), blackTile )
```

Comment:
I think the intention of this block was just to check that the images aren't black, so that the more detailed assertions below are watertight. I think we'd be better off moving the `fileValid` checks into `assertBlack()`, `assertMatch()` etc. And I'm not sure the current behaviour is what we want - when `ClampToFrame` takes effect I think we want `fileValid` to be `true`, and I _think_ when `BlackOutside` takes effect we want `fileValid` to be `false`. In other words, I think we want `fileValid` to mean "you can see pixels in the viewer and they came from a real file". Does that make sense? Any thoughts, @murraystevenson?

Reply:
Ah, good point, let me update the tests to define that behaviour

Reply:
have updated the logic to work with `missingFrameMode` which in turn works with the frame mask modes. Have updated the unit tests so that each of the modes are now tested.